File: src/rhx/providers/crypto_official.py
from __future__ import annotations

import base64
import json
import time
import uuid
from typing import Any

import httpx
from nacl.signing import SigningKey

from rhx.auth import AuthManager
from rhx.errors import CLIError, ErrorCode
from rhx.models import (
    AuthStatus,
    CapabilitySet,
    OrderIntentCrypto,
    OrderResult,
)
from rhx.providers.base import OrderIntent


class RobinhoodCryptoProvider:
    name = "crypto"

    def __init__(self, auth: AuthManager, base_url: str = "https://trading.robinhood.com") -> None:
        self.auth = auth
        self.base_url = base_url.rstrip("/")

    def capabilities(self) -> CapabilitySet:
        return CapabilitySet(stocks=False, crypto=True, options=False, options_spreads=False)

    def _credentials(self) -> tuple[str, str]:
        api_key, private_key_b64 = self.auth.crypto_credentials()
        if not api_key or not private_key_b64:
            raise CLIError(
                code=ErrorCode.AUTH_REQUIRED,
                message="Missing RH_CRYPTO_API_KEY or RH_CRYPTO_PRIVATE_KEY_B64",
            )
        return api_key, private_key_b64

    def _sign(self, private_key_b64: str, message: str) -> str:
        try:
            key_bytes = base64.b64decode(private_key_b64)
        except Exception as exc:
            raise CLIError(code=ErrorCode.AUTH_REQUIRED, message=f"Invalid crypto private key encoding: {exc}") from exc

        # Ed25519 private key material should be a 32-byte seed or 64-byte secret/public concat.
        if len(key_bytes) not in (32, 64):
            raise CLIError(
                code=ErrorCode.AUTH_REQUIRED,
                message="Invalid crypto private key length; expected 32 or 64 decoded bytes",
            )

        seed = key_bytes[:32]
        signing_key = SigningKey(seed)
        signature = signing_key.sign(message.encode("utf-8")).signature
        return base64.b64encode(signature).decode("utf-8")

    def _request(
        self,
        method: str,
        path: str,
        *,
        payload: dict[str, Any] | None = None,
        params: dict[str, Any] | None = None,
    ) -> Any:
        api_key, private_key_b64 = self._credentials()
        timestamp = str(int(time.time()))
        body = json.dumps(payload, separators=(",", ":"), sort_keys=True) if payload is not None else ""

        signing_payload = f"{api_key}{timestamp}{path}{method.upper()}{body}"
        signature = self._sign(private_key_b64, signing_payload)

        headers = {
            "x-api-key": api_key,
            "x-signature": signature,
            "x-timestamp": timestamp,
        }
        if payload is not None:
            headers["Content-Type"] = "application/json"

        url = f"{self.base_url}{path}"
        with httpx.Client(timeout=20.0) as client:
            response = client.request(method, url, headers=headers, json=payload, params=params)

        if response.status_code == 429:
            raise CLIError(code=ErrorCode.RATE_LIMITED, message="Robinhood crypto API rate limit", retriable=True)
        if response.status_code in (401, 403):
            raise CLIError(code=ErrorCode.AUTH_REQUIRED, message=f"Crypto auth failed: {response.text}")
        if response.status_code >= 400:
            raise CLIError(code=ErrorCode.BROKER_REJECTED, message=f"Crypto API error {response.status_code}: {response.text}")

        if response.content:
            try:
                return response.json()
            except Exception:
                return {"text": response.text}
        return {}

    def auth_status(self) -> AuthStatus:
        try:
            self._request("GET", "/api/v1/crypto/trading/accounts/")
            return AuthStatus(provider="crypto", authenticated=True, detail="Authenticated")
        except CLIError as exc:
            return AuthStatus(
                provider="crypto",
                authenticated=False,
                mfa_required=False,
                detail=exc.message,
            )

    def account_summary(self) -> dict[str, Any]:
        data = self._request("GET", "/api/v1/crypto/trading/accounts/")
        return data if isinstance(data, dict) else {"accounts": data}

    def positions(self) -> list[dict[str, Any]]:
        data = self._request("GET", "/api/v1/crypto/trading/holdings/")
        if isinstance(data, dict):
            for key in ("results", "holdings"):
                if isinstance(data.get(key), list):
                    return data[key]
            return [data]
        return data if isinstance(data, list) else []

    def quote(self, symbol: str) -> dict[str, Any]:
        data = self._request(
            "GET",
            "/api/v1/crypto/marketdata/best_bid_ask/",
            params={"symbol": symbol},
        )
        return {"symbol": symbol, "quote": data}

    def place_order(self, intent: OrderIntent) -> OrderResult:
        if not isinstance(intent, OrderIntentCrypto):
            raise CLIError(code=ErrorCode.VALIDATION_ERROR, message="Crypto provider only supports crypto orders")

        order_type = intent.order_type
        amount_in = intent.amount_in

        payload: dict[str, Any] = {
            "client_order_id": str(uuid.uuid4()),
            "side": intent.side,
            "symbol": intent.symbol,
            "type": order_type,
            "time_in_force": intent.time_in_force,
        }

        if order_type == "market":
            if amount_in == "quantity":
                payload["market_order_config"] = {"asset_quantity": str(intent.quantity)}
            else:
                payload["market_order_config"] = {"quote_amount": str(intent.notional_usd)}
        else:
            if amount_in == "quantity":
                payload["limit_order_config"] = {
                    "asset_quantity": str(intent.quantity),
                    "limit_price": str(intent.limit_price),
                }
            else:
                payload["limit_order_config"] = {
                    "quote_amount": str(intent.notional_usd),
                    "limit_price": str(intent.limit_price),
                }

        raw = self._request("POST", "/api/v1/crypto/trading/orders/", payload=payload)

        if isinstance(raw, dict):
            order_id = raw.get("id") or raw.get("order_id")
            state = raw.get("state") or raw.get("status")
        else:
            order_id = None
            state = None

        return OrderResult(
            provider=self.name,
            order_id=order_id,
            state=state,
            symbol=intent.symbol,
            side=intent.side,
            asset_type="crypto",
            raw=raw if isinstance(raw, dict) else {"response": raw},
        )

    def cancel_order(self, order_id: str, asset_type: str | None = None) -> dict[str, Any]:
        data = self._request("POST", f"/api/v1/crypto/trading/orders/{order_id}/cancel/")
        return {"asset_type": "crypto", "result": data}

    def get_order(self, order_id: str, asset_type: str | None = None) -> dict[str, Any]:
        data = self._request("GET", f"/api/v1/crypto/trading/orders/{order_id}/")
        return {"asset_type": "crypto", "order": data}

    def list_orders(self, open_only: bool = False, asset_type: str | None = None) -> list[dict[str, Any]]:
        params = {"state": "open"} if open_only else None
        data = self._request("GET", "/api/v1/crypto/trading/orders/", params=params)

        if isinstance(data, dict):
            if isinstance(data.get("results"), list):
                return data["results"]
            return [data]
        return data if isinstance(data, list) else []

File: pkg/rhx/crypto_official.go
package rhx

import (
	"bytes"
	"context"
	"crypto/ed25519"
	"encoding/base64"
	"encoding/json"
	"net/http"
	"strconv"
	"strings"
	"time"
)

type OfficialCryptoProvider struct {
	auth *AuthManager
	base string
}

type CryptoOrderIntent struct {
	Symbol      string
	Side        string
	Type        string
	AmountIn    string
	Quantity    *float64
	NotionalUSD *float64
	LimitPrice  *float64
	TimeInForce string
}

func newOfficialCryptoProvider(auth *AuthManager) *OfficialCryptoProvider {
	return &OfficialCryptoProvider{auth: auth, base: robinhoodTradingBase}
}

func (p *OfficialCryptoProvider) credentials() (string, string, error) {
	apiKey, privateKey, _ := p.auth.Store.cryptoCredentials(p.auth.Profile)
	if apiKey == "" || privateKey == "" {
		return "", "", newError(ErrorAuthRequired, "Missing RH_CRYPTO_API_KEY or RH_CRYPTO_PRIVATE_KEY_B64")
	}
	return apiKey, privateKey, nil
}

func (p *OfficialCryptoProvider) verify(ctx context.Context) error {
	_, err := p.request(ctx, http.MethodGet, "/api/v1/crypto/trading/accounts/", nil, nil)
	return err
}

func (p *OfficialCryptoProvider) accountSummary(ctx context.Context) (map[string]any, error) {
	data, err := p.request(ctx, http.MethodGet, "/api/v1/crypto/trading/accounts/", nil, nil)
	if err != nil {
		return nil, err
	}
	return asMap(data), nil
}

func (p *OfficialCryptoProvider) positions(ctx context.Context) ([]map[string]any, error) {
	data, err := p.request(ctx, http.MethodGet, "/api/v1/crypto/trading/holdings/", nil, nil)
	if err != nil {
		return nil, err
	}
	rows := resultsRows(data)
	if len(rows) == 0 {
		if m := asMap(data); len(m) > 0 {
			rows = []map[string]any{m}
		}
	}
	return rows, nil
}

func (p *OfficialCryptoProvider) quote(ctx context.Context, symbol string) (map[string]any, error) {
	data, err := p.request(ctx, http.MethodGet, "/api/v1/crypto/marketdata/best_bid_ask/", map[string]string{"symbol": normalizeCryptoSymbol(symbol)}, nil)
	if err != nil {
		return nil, err
	}
	return map[string]any{"symbol": normalizeCryptoSymbol(symbol), "quote": data}, nil
}

func (p *OfficialCryptoProvider) quotes(ctx context.Context, symbols []string) ([]map[string]any, error) {
	rows := []map[string]any{}
	for _, symbol := range symbols {
		row, err := p.quote(ctx, symbol)
		if err != nil {
			return nil, err
		}
		rows = append(rows, row)
	}
	return rows, nil
}

func (p *OfficialCryptoProvider) placeOrder(ctx context.Context, intent CryptoOrderIntent) (map[string]any, float64, error) {
	payload := map[string]any{
		"client_order_id": randomDeviceToken(),
		"side":            intent.Side,
		"symbol":          normalizeCryptoSymbol(intent.Symbol),
		"type":            intent.Type,
		"time_in_force":   intent.TimeInForce,
	}
	estimated := 0.0
	if intent.Type == "market" {
		if intent.AmountIn == "quantity" {
			if intent.Quantity == nil {
				return nil, 0, newError(ErrorValidation, "--qty is required when --amount-in quantity")
			}
			payload["market_order_config"] = map[string]any{"asset_quantity": formatFloat(*intent.Quantity)}
		} else {
			if intent.NotionalUSD == nil {
				return nil, 0, newError(ErrorValidation, "--notional-usd is required when --amount-in price")
			}
			estimated = *intent.NotionalUSD
			payload["market_order_config"] = map[string]any{"quote_amount": formatFloat(*intent.NotionalUSD)}
		}
	} else {
		if intent.LimitPrice == nil {
			return nil, 0, newError(ErrorValidation, "--limit-price is required for limit orders")
		}
		if intent.AmountIn == "quantity" {
			if intent.Quantity == nil {
				return nil, 0, newError(ErrorValidation, "--qty is required when --amount-in quantity")
			}
			estimated = *intent.Quantity * *intent.LimitPrice
			payload["limit_order_config"] = map[string]any{"asset_quantity": formatFloat(*intent.Quantity), "limit_price": formatFloat(*intent.LimitPrice)}
		} else {
			if intent.NotionalUSD == nil {
				return nil, 0, newError(ErrorValidation, "--notional-usd is required when --amount-in price")
			}
			estimated = *intent.NotionalUSD
			payload["limit_order_config"] = map[string]any{"quote_amount": formatFloat(*intent.NotionalUSD), "limit_price": formatFloat(*intent.LimitPrice)}
		}
	}
	data, err := p.request(ctx, http.MethodPost, "/api/v1/crypto/trading/orders/", nil, payload)
	if err != nil {
		return nil, 0, err
	}
	raw := asMap(data)
	return map[string]any{
		"provider":   "crypto",
		"order_id":   firstString(raw, "id", "order_id"),
		"state":      firstString(raw, "state", "status"),
		"symbol":     normalizeCryptoSymbol(intent.Symbol),
		"side":       intent.Side,
		"asset_type": "crypto",
		"raw":        raw,
	}, estimated, nil
}

func (p *OfficialCryptoProvider) listOrders(ctx context.Context, openOnly bool) ([]map[string]any, error) {
	query := map[string]string(nil)
	if openOnly {
		query = map[string]string{"state": "open"}
	}
	data, err := p.request(ctx, http.MethodGet, "/api/v1/crypto/trading/orders/", query, nil)
	if err != nil {
		return nil, err
	}
	return resultsRows(data), nil
}

func (p *OfficialCryptoProvider) getOrder(ctx context.Context, orderID string) (map[string]any, error) {
	data, err := p.request(ctx, http.MethodGet, "/api/v1/crypto/trading/orders/"+orderID+"/", nil, nil)
	if err != nil {
		return nil, err
	}
	return map[string]any{"asset_type": "crypto", "order": asMap(data)}, nil
}

func (p *OfficialCryptoProvider) cancelOrder(ctx context.Context, orderID string) (map[string]any, error) {
	data, err := p.request(ctx, http.MethodPost, "/api/v1/crypto/trading/orders/"+orderID+"/cancel/", nil, map[string]any{})
	if err != nil {
		return nil, err
	}
	return map[string]any{"asset_type": "crypto", "result": data}, nil
}

func (p *OfficialCryptoProvider) request(ctx context.Context, method string, path string, query map[string]string, payload any) (any, error) {
	apiKey, privateKeyB64, err := p.credentials()
	if err != nil {
		return nil, err
	}
	body := ""
	var reader *bytes.Reader
	if payload != nil {
		b, err := json.Marshal(payload)
		if err != nil {
			return nil, err
		}
		body = string(b)
		reader = bytes.NewReader(b)
	} else {
		reader = bytes.NewReader(nil)
	}
	timestamp := strconvFormatUnix(time.Now().Unix())
	signingPayload := apiKey + timestamp + path + strings.ToUpper(method) + body
	signature, err := signEd25519(privateKeyB64, signingPayload)
	if err != nil {
		return nil, err
	}
	req, err := http.NewRequestWithContext(ctx, method, p.base+path, reader)
	if err != nil {
		return nil, err
	}
	if query != nil {
		q := req.URL.Query()
		for key, value := range query {
			q.Set(key, value)
		}
		req.URL.RawQuery = q.Encode()
	}
	req.Header.Set("x-api-key", apiKey)
	req.Header.Set("x-signature", signature)
	req.Header.Set("x-timestamp", timestamp)
	req.Header.Set("Accept", "application/json")
	if payload != nil {
		req.Header.Set("Content-Type", "application/json")
	}
	data, _, err := newHTTPClient(nil).do(req)
	return data, err
}

func signEd25519(privateKeyB64 string, message string) (string, error) {
	keyBytes, err := base64.StdEncoding.DecodeString(privateKeyB64)
	if err != nil {
		return "", wrapError(ErrorAuthRequired, "Invalid crypto private key encoding: %v", err)
	}
	var privateKey ed25519.PrivateKey
	switch len(keyBytes) {
	case ed25519.SeedSize:
		privateKey = ed25519.NewKeyFromSeed(keyBytes)
	case ed25519.PrivateKeySize:
		privateKey = ed25519.PrivateKey(keyBytes)
	default:
		return "", newError(ErrorAuthRequired, "Invalid crypto private key length; expected 32 or 64 decoded bytes")
	}
	sig := ed25519.Sign(privateKey, []byte(message))
	return base64.StdEncoding.EncodeToString(sig), nil
}

func strconvFormatUnix(ts int64) string {
	return strconv.FormatInt(ts, 10)
}
